contracts/Marketplace.sol

```diff
@@ -359,176 +338,96 @@ contract Marketplace is SlotReservations, Proofs, StateRetrieval, Endian {
-     deposited them.
-   * @dev Request must be cancelled, failed or finished, and the
-     transaction must originate from the depositor address.
-   * @param requestId the id of the request
-   */
-  function withdrawFunds(RequestId requestId) public {
-    withdrawFunds(requestId, msg.sender);
-  }
-
-  /**
-     * @notice Withdraws storage request funds to the provided address.
-   * @dev Request must be expired, must be in RequestState.New, and the
-     transaction must originate from the depositer address.
-   * @param requestId the id of the request
-   * @param withdrawRecipient address to return the remaining funds to
-   */
-  function withdrawFunds(
-    RequestId requestId,
-    address withdrawRecipient
-  ) public requestIsKnown(requestId) {
-    Request storage request = _requests[requestId];
-    RequestContext storage context = _requestContexts[requestId];
-
-    if (request.client != msg.sender) revert Marketplace_InvalidClientAddress();
+  /// Withdraws remaining storage request funds back to the client that
+  function withdrawFunds(RequestId requestId) public requestIsKnown(requestId) {
+    FundId fund = requestId.asFundId();
+    AccountId account = _vault.clientAccount(msg.sender);
+    _vault.withdraw(fund, account);
 
-    RequestState state = requestState(requestId);
-    if (
-      state != RequestState.Cancelled &&
-      state != RequestState.Failed &&
-      state != RequestState.Finished
-    ) {
-      revert Marketplace_InvalidState();
-    }
-
-    // fundsToReturnToClient == 0 is used for "double-spend" protection, once the funds are withdrawn
-    // then this variable is set to 0.
-    if (context.fundsToReturnToClient == 0)
-      revert Marketplace_NothingToWithdraw();
-
-    if (state == RequestState.Cancelled) {
-      context.state = RequestState.Cancelled;
-      emit RequestCancelled(requestId);
-
-      // `fundsToReturnToClient` currently tracks funds to be returned for requests that successfully finish.
-      // When requests are cancelled, funds earmarked for payment for the duration
-      // between request expiry and request end (for every slot that was filled), should be returned to the client.
-      // Update `fundsToReturnToClient` to reflect this.
-      context.fundsToReturnToClient +=
-        context.slotsFilled *
-        _slotPayout(requestId, requestExpiry(requestId));
-    } else if (state == RequestState.Failed) {
-      // For Failed requests the client is refunded whole amount.
-      context.fundsToReturnToClient = request.maxPrice();
-    } else {
-      context.state = RequestState.Finished;
-    }
-
-    _removeFromMyRequests(request.client, requestId);
-
-    uint256 amount = context.fundsToReturnToClient;
-    _marketplaceTotals.sent += amount;
-
-    if (!_token.transfer(withdrawRecipient, amount)) {
-      revert Marketplace_TransferFailed();
-    }
+    _removeFromMyRequests(msg.sender, requestId);
```

Comment:
This function does not perform any verifications that the caller is indeed client of the request. 
Technically it should not really be a problem as it will withdraw whatever the caller has on the Vault accounts, it will not withdraw client's funds. But you are effectively exposing the Vault's `withdraw` function through this. Maybe this could be simplified with having one `withdraw` function which would accept the `VaultRole` enum in order to withdraw from the correct account?

Reply:
Good point, I should probably re-add the check that `request.client == msg.sender`.

> Maybe this could be simplified with having one `withdraw` function which would accept the `VaultRole` enum in order to withdraw from the correct account?

I tried to avoid adding VaultRole to the API and leave it as an implementation detail.

Reply:
Make sense 👍 It was more note to understand if it is a problem or not - I agree it is not.